### PR TITLE
Deviceplugin refactoring: merge func list and listwatch in endpoint into one

### LINKED ISSUE #51993: Device plugin code cleanup
/kind feature

Here are the possible code cleanup/refactoring we have discussed during device plugin PR code review or offline:
1. Currently Kubelet side device plugin is structured as three-layers: cm.DevicePluginHandler -> deviceplugin.Manager -> deviceplugin.endpoint. We may improve code readability by dropping them to two: cm.DevicePluginHandler -> deviceplugin.Endpoint. Some of the functionalities of deviceplugin.Manager will be moved to cm.DevicePluginHandler and some may be moved to deviceplugin.Endpoint.
2. The current deviceplugin.Endpoint has two methods: list and listAndWatch that make the same gRPC call. We may merge them into one. Combined with 1) item, the deviceplugin.Endpoint interface may expose the following methods: Start(), Stop(), Alloc(), run().
- Interface cleanup: some of the interfaces in deviceplugin pkg is not necessary public and can be changed to internal.
- Most functions in deviceplugin/utils.go doesn't seem to be necessary. May consider to remove this file.

### PR DESCRIPTION
**Pull request by lichuqiang**
**What this PR does / why we need it**:
merge func list and listwatch in endpoint into one, since we won't call list func individually

**Which issue this PR fixes**
fixes #51993
Part2

**Special notes for your reviewer**:
/cc @jiayingz @RenaudWasTaken @vishh

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Review by RenaudWasTaken — COMMENTED** [review]
The behavior here is different from the previous implementation.
It's actually a good way of removing the whole list function.

An idea that comes to mind here (probably a follow-up PR) would be to un-register a device plugin that does not register any devices within 5 seconds of registration.

**Comment by lichuqiang** [comment]
@RenaudWasTaken I've tried to update the register timeout check in current PR, you can take a look.
I prefer to add the check in manager. But, this way, I'm afraid I have to add another goroutine for the check loop, which may impact on overall performance.
Another way is to add check directly to `listAndWatch` func of endpoint, but I think it looks weird this way.
Do you have any suggestions?

**Comment by RenaudWasTaken** [comment]
> @RenaudWasTaken I've tried to update the register timeout check in current PR, you can take a look.

I was thinking that you might add that in a separate PR :)

> I prefer to add the check in manager.

I think it's a good idea

> But, this way, I'm afraid I have to add another goroutine for the check loop, which may impact on overall performance.

That's the idiomtic way to go from what I understand. I wouldn't be concerned over performance, goroutines are [pretty cheap](http://blog.nindalf.com/how-goroutines-work/) and this one isn't a long running one so it won't impact performance. 

> Another way is to add check directly to listAndWatch func of endpoint, but I think it looks weird this way.

Agreed

> Do you have any suggestions?

I've added a few of them :)

File: pkg/kubelet/deviceplugin/manager.go
Comment (RenaudWasTaken) [inline pkg/kubelet/deviceplugin/manager.go]: I usually use the pattern:

```
if !condition1 {
     return
}
if !condition2 {
     return
}
action()
}
```

Instead of the pattern:
```
if condition1 {
    if condition2 {
         action()
    }
}
if condition2 {
     return
}
action
}
```

It makes the code less indented which feels easier to read and makes it easier later to add `if !condition3`

File: pkg/kubelet/deviceplugin/manager.go
Comment (RenaudWasTaken) [inline pkg/kubelet/deviceplugin/manager.go]: If you defer the mutex Unlock right after the lock you won't have to unlock it twice.
It's also usually more idiomatic :)

File: pkg/kubelet/deviceplugin/manager.go
Comment (RenaudWasTaken) [inline pkg/kubelet/deviceplugin/manager.go]: I think you should just wait for the time to tick and check at the end if there are devices or not that would remove the need for a channel

Comment (RenaudWasTaken) [reply pkg/kubelet/deviceplugin/manager.go]: How about:
```
<-time.Affter(time.Second * 2)
```

File: pkg/kubelet/deviceplugin/manager.go
Comment (lichuqiang) [reply pkg/kubelet/deviceplugin/manager.go]: good catch

Comment (lichuqiang) [reply pkg/kubelet/deviceplugin/manager.go]: good catch

File: pkg/kubelet/deviceplugin/manager.go
Comment (lichuqiang) [reply pkg/kubelet/deviceplugin/manager.go]: But this way, it would block until timeout, and we could not handle in time if listWatch of endpoint exits during this period

Comment (RenaudWasTaken) [reply pkg/kubelet/deviceplugin/manager.go]: I was still commenting in the context of the goroutine :)
Thought about this a bit more here is what I think might be a cleaner solution.

```golang
go func() {
	<-time.After(5 * time.Second)

	e.mutex.Lock()
	nDevices = len(e.devices)
	e.mutex.Unlock()

	if nDevices == 0 {
		e.stop()
	}

	select {
		case <-e.stopChan:
			e.stop()
		case <-e.exitChan:
			removeEndpoint(e)
			return
	}
}

func (e *endpoint) stop() {
	e.cancel()
}
```

And you would change Endpoint.runlike this:
```
func (e *endpoint) run() {
     defer close(e.exitChan)
}```

Comment (RenaudWasTaken) [reply pkg/kubelet/deviceplugin/manager.go]: The snippet is missing a for loop before the select

File: pkg/kubelet/deviceplugin/manager.go
Comment (jiayingz) [inline pkg/kubelet/deviceplugin/manager.go]: @lichuqiang, could you take a look at my PR https://github.com/kubernetes/kubernetes/pull/52290 that is to fix the race condition during re-registration? A unit test in manager-test.go is added in that PR for this. The problem is also exposed through the PR that extended e2e gpu test to exercise device plugin restarts: https://github.com/kubernetes/kubernetes/pull/52375.

Comment (lichuqiang) [reply pkg/kubelet/deviceplugin/manager.go]: Yes, I see your PR, a good catch, to test is always a good way to spot problems:)
but I don't think the problem has relationship with changes here, it's mostly caused by the lack of callback and corresponding update in node capacity when removing endpoint, as you fixed in your PR. And the change @RenaudWasTaken suggested here is just to handle rededundant plugin that doesn't register devices indeed.
I would prefer revert the PR to it's original state just for the combination of func list & listwatch, and maybe open another PR about the registration issue if necessary, thanks.

Comment (jiayingz) [reply pkg/kubelet/deviceplugin/manager.go]: +1 for separating them into different PRs.

File: pkg/kubelet/deviceplugin/manager.go
Comment (RenaudWasTaken) [inline pkg/kubelet/deviceplugin/manager.go]: This will re-open issue #52510

**Review by RenaudWasTaken — CHANGES_REQUESTED** [review]

File: pkg/kubelet/deviceplugin/manager.go
Comment (RenaudWasTaken) [reply pkg/kubelet/deviceplugin/manager.go]: So I was thinking the following might be a solution:

```golang
// run guarantees that if error is not nil, it has received at least one call to ListAndWatch by the client Device Plugin 
func (e *endpoint) run() error {
	callback := e.callback
	deviceChan := make(chan interface{})

	e.callback = func(n string, a, u, r []*pluginapi.Device) {
		callback(n, a, u, r)
		e.callback = callback

		close(deviceChan)
	}
	
	go e.listAndWatch()
	<-deviceChan

     return nil
}
```

Comment (lichuqiang) [reply pkg/kubelet/deviceplugin/manager.go]: Well, I'm not sure about your meaning. Is it in order to prevent that the plugin newly registered could not listAndWatch successfully at all, so that the the old devices in handler & node capacity could not be deleted correspondingly?
I don't think the change you suggested could take effect on it. How would it behavior if an error occurs? Lead to a block?
How about move `e.callback(e.resourceName, []*pluginapi.Device{}, []*pluginapi.Device{}, e.getDevices())` into `e.stop()`? since the stop func is thoutht to be a cleanup func of endpoint

**Comment by RenaudWasTaken** [comment]
Hi @lichuqiang!

Sorry for the very late answer. I'd like to see this PR merged ASAP.
I'm re-reading this PR and code and here is my general feeling:

I'm not a big fan of adding an `endpoint` in the `m.endpoints` map if there are no devices associated to this endpoint.
There are two reasons:
- This means that a "spammy" device plugin (one that re-registers multiple times) can impact the registration of other device plugins
- It seems more logical, if we are to have a timeout feature on registering devices, that we never add the endpoint in the map in the first place rather than adding it, waiting for the timeout and then removing it.

Which is why I was suggesting that the `run` method returns only when devices have been registered and starts a go routine see:
```golang
// run starts a go routine that listens for device changes
// run guarantees that if error is not nil, it has received at least one
// call to ListAndWatch by the client Device Plugin 
func (e *endpoint) run() error {
	callback := e.callback
	deviceChan := make(chan interface{})

	e.callback = func(n string, a, u, r []*pluginapi.Device) {
		callback(n, a, u, r)
		e.callback = callback

		close(deviceChan)
	}
	
	go e.listAndWatch()
	<-deviceChan

     return nil
}
```

with manager.go:
```
func (m *ManagerImpl) addEndpoint(r *pluginapi.RegisterRequest) {
// ...
    err := e.run()
    if err != nil {
        glog.Errorf(...)
        return
    }

    m.mutex.Lock()
    old, ok := m.endpoints[r.ResourceName]
    m.endpoints[r.ResourceName] = e
    m.mutex.Unlock()
// ...
}
```

@jiayingz @lichuqiang does that seem reasonable ?

**Comment by lichuqiang** [comment]
@RenaudWasTaken Seems func run() would keep running if no devices reported. Thus, could the resources consumed by the goroutine get recycled?

**Comment by RenaudWasTaken** [comment]
> @RenaudWasTaken Seems func run() would never return if no devices reported. Thus, could the resources consumed by the goroutine get recycled?

This is the current behavior of `list` so no real changes here.
I'd like to get a timeout in a follow-up PR, hoping to keep the scope of this PR as small as possible

**Comment by jiayingz** [comment]
I would suggest we avoid introducing timeout in this PR.

There is however some corner case to consider. First, you may want to take a look at TestDevicePluginReRegistration(). I think that unit test was added after this PR was implemented and would fail in its current format. That unit test is currently flaky anyway. We may change it to cope with the changed behavior, i.e., a new callback will always be issued with the new endpoint. However, there is a real issue that we may fail to remove some devices that happen to be removed after a new device plugin is deployed. Consider the scenario that the first device plugin reports device 1, 2, 3 after its registration. Then we do a device plugin upgrade. Say the new device plugin only reports device 1 and 2. With the current code, we will never remove device 3. Note that this problem exists in the current code as well.

To fix this problem, one possible solution is to pass list of current devices in newEndpoint(). For the first registered device plugin, the list will be empty. For the subsequent re-registered plugin, the list will be what reported by its previous device plugin.

**Comment by RenaudWasTaken** [comment]
> I think that unit test was added after this PR was implemented and would fail in its current format

That's not how the CI works, it rebases the PR over master at each runs.

> We may change it to cope with the changed behavior, i.e., a new callback will always be issued with the new endpoint

A new callback isn't issued with this refactor.

> However, there is a real issue that we may fail to remove some devices that happen to be removed after a new device plugin is deployed. 

I don't think this is the scope of this PR and should be given another PR. Do you want to open an issue for this @jiayingz ?

**Comment by jiayingz** [comment]
/lgtm

I am fine to address this in a separate PR.

**Comment by RenaudWasTaken** [comment]
@jiayingz I still think this PR should be changed to address the following issues:

- Current implementation means that an endpoint will get added in the endpoint map after registration even if it doesn't call `listWatch`.
   This means that if one endpoint spams registration this will impact registration for other endpoints, we could at least keep the current behavior of requiring one call to `listWatch` before adding the endpoint to the map.

- Furthermore the model introduced will probably be changed so that we can add a clean timeout afterwards.
   I'd rather we keep the current (master) model rather than change it to this PR model and then revert it afterwards (which seems silly).

It doesn't bother me to introduce timeout in a follow-up PR but i'd rather we start introducing a more robust and modular model now which will allow us to build:
- spam prevention
- timeout
- start thinking about registration and kubelet restart

**Comment by jiayingz** [comment]
I think it is ok to add an endpoint before calling listwatch. If we are worried abut spam device plugin, it could happen through a flood of listwatch as well.

We do have a TODO item on how to prevent spam device plugins, but I think that is beyond the scope of this PR. I also hope the solution we come up can prevent spam listwatch as well as spam registration.

If we do want to change the scope of this PR, personally, I think passing device list from previous device plugin takes higher priority as it could affect correctness and would also help mitigate the listwatch delay problem.

**Comment by RenaudWasTaken** [comment]
> I think it is ok to add an endpoint before calling listwatch. If we are worried abut spam device plugin, it could happen through a flood of listwatch as well
> We do have a TODO item on how to prevent spam device plugins, but I think that is beyond the scope of this PR. I also hope the solution we come up can prevent spam listwatch as well as spam registration.

Keeping the same behavior seems to be in the scope of this PR though.
Agreed that preventing spam is beyond the scope of this PR, however allowing spam from re-registration (multiple device plugin, restart, crash, ...) seems to be something we don't want to re-introduce.

Also since we are actually changing both function we might as well discuss and implement the architecture that allows timeout and spam prevention in this PR.

**Comment by jiayingz** [comment]
I still feel that requiring a device plugin to have a successful listwatch before being added as an endpoint itself is not enough to prevent device plugin spam anyway, given that generally a device plugin will send listwatch immediately after its registration, and can crash after that. To prevent device plugin spam, we probably want to track no. of different grpc updates during a time period and implement some de-prioritizing policy for too frequent grpc updates (including registration as well as listwatch). I feel this is beyond the scope of this PR. Given the current trust model of device plugin, I also feel implementing de-prioritizing policy is probably lower priority but tracking no. of different grpc updates during a time period may be more important for monitoring purpose.

**Comment by RenaudWasTaken** [comment]
> I still feel that requiring a device plugin to have a successful listwatch before being added as an endpoint itself is not enough to prevent device plugin spam anyway, given that generally a device plugin will send listwatch immediately after its registration, and can crash after that. 

I don't think it's a great argument to say that since we can always spam the system, we might as well drop all the spam prevention we have.
Some spam prevention is still better than none at all. Especially if the prevention is already in place.
Let's at least keep this behavior.

> Given the current trust model of device plugin, I also feel implementing de-prioritizing policy is probably lower priority but tracking no. of different grpc updates during a time period may be more important for monitoring purpose.

Let's open an issue for this.

> To prevent device plugin spam, we probably want to track no. of different grpc updates during a time period and implement some de-prioritizing policy for too frequent grpc updates (including registration as well as listwatch). I feel this is beyond the scope of this PR. 

Agreed which is why spam prevention is not my main point.

My main point revolves around keeping the current behavior and laying the grounds to the next architecture (timeout, spam prevention).

Which I think the code snippet mentioned above should suffice.

IMHO far

**Comment by jiayingz** [comment]
If the described spam scenario is a common scenario, I agree we should keep its handling logic. However, I don't think we have any data to suggest that at this point. Also I don't think the mentioned code snippet itself can prevent any spam, given that the endpoint is already added before calling its run function.

**Comment by tengqm** [comment]
This needs are base now ..

**Comment by RenaudWasTaken** [comment]
> If the described spam scenario is a common scenario, I agree we should keep its handling logic. However, I don't think we have any data to suggest that at this point.

As mentioned dropping all spam prevention because we don't see any spam doesn't sound like a good idea. Requiring a call to ListAndWatch might not be the best prevention but it's still better than nothing.

> Also I don't think the mentioned code snippet itself can prevent any spam, given that the endpoint is already added before calling its run function.

See the code snippet just under the one mentioned that describes the manager change.

See [this design document](https://docs.google.com/document/d/1jevkhlJ--GGdBt2Etyrd8mrQfs9QMk8GgHa9nzpJR1M) for more information on how I see the endpoint.

**Comment by jiayingz** [comment]
@RenaudWasTaken It is still hard for me to see the benefit of the posted code snippet. Whether we return an error in endpoint.run() or not, it takes similar amount of processing on the kubelet side, so I don't see how this helps prevent spam. Perhaps I missed some point here, but given that this doesn't seem to be a straightforward addition on top of this PR, I would prefer to not including it in this PR.

**Comment by RenaudWasTaken** [comment]
> @RenaudWasTaken It is still hard for me to see the benefit of the posted code snippet. Whether we return an error in endpoint.run() or not, it takes similar amount of processing on the kubelet side, so I don't see how this helps prevent spam.  Perhaps I missed some point here, but given that this doesn't seem to be a straightforward addition on top of this PR, I would prefer to not including it in this PR.

The current behavior and the one I am trying to keep is that we require a call to `ListAndWatch` before adding the endpoint to the mutex.
Which means that to actually spam the mutex lock and impact other plugins it becomes a lot more complex than just a single for loop.

It's actually pretty complexe with the described behavior to impact registration of other device plugins.
With the current behavior, a device plugin with a gRPC call looping or even just crashing could prevent any possible registration.

Furthermore the mentioned code snippet is a first step in the Endpoint being in charge of it's own lifecycle and thus the life and death of the associated goroutine.

In terms of code I am going to re-post the `manager` code that I've put higher. It's closer to the current implementation than this PR's one.

```golang
func (m *ManagerImpl) addEndpoint(r *pluginapi.RegisterRequest) {
// ...
    if err := e.run(); err != nil {
        glog.Errorf(...)
        return
    }

    m.mutex.Lock()
    old, ok := m.endpoints[r.ResourceName]
    m.endpoints[r.ResourceName] = e
    m.mutex.Unlock()
// ...
}```

**Comment by jiayingz** [comment]
Sorry I just saw your recent code snippet but missed the one above it. Looking at that code snippet, I still don't think it would work as is. First, with this PR, we planned to remove endpoint listAndWatch() but I guess you propose to still keep it and wrap it with endpoint.run()? Even with timeout, I think we still need to worry about go routine leaks on the one used for listAndWatch(). We may say that we close endpoint connection upon timeout so that listAndWatch() go routine would fail and exit, but I also have concerns on requiring an explicit time bound between registration and listWatch. Often after we introduce a timeout, we end up adjusting it from time to time to cope with various situations. Also the requirement to finish the first listWatch before adding the endpoint could also delay the recovery after kubelet restart that currently require all device plugins to re-register for new device allocation requests. Given these open questions, I think it is better address the spam detection/prevention problem in a separate PR.

**Comment by RenaudWasTaken** [comment]
> Sorry I just saw your recent code snippet but missed the one above it. Looking at that code snippet, I still don't think it would work as is. 

Can you list actual pain points ?

> First, with this PR, we planned to remove endpoint listAndWatch() but I guess you propose to still keep it and wrap it with endpoint.run()? 

Yes and still merge list and listWatch

> I think we still need to worry about go routine leaks on the one used for listAndWatch(). 

I'm not sure what you are mentioning here ?
As I have developped in the design document, my intent is to have the endpoint in charge of it's own goroutines and lifecycle, not the manager.
The current model is painful and not adaptable to change. Interactions between the endpoint and the manager are pretty much un-intuitive as the endpoint has at least 2 ways to be closed.

Though my intent is not to introduce all these changes in this PR, it would at least be a first step in the direction to have a saner architecture.

> We may say that we close endpoint connection upon timeout so that listAndWatch() go routine would fail and exit, but I also have concerns on requiring an explicit time bound between registration and listWatch.

We don't have to implement this in this PR.

> Often after we introduce a timeout, we end up adjusting it from time to time to cope with various situations. 

Let's not get into this conversation just yet. However please note that the expected lifecycle of a device plugin is that we expect the device plugin to be ready to advertise devices when registering.

> Also the requirement to finish the first listWatch before adding the endpoint could also delay the recovery after kubelet restart that currently require all device plugins to re-register for new device allocation requests. 

As mentioned higher the lifecycle of the device plugin is that we expect the device plugin to be ready to advertise devices when registering.
Furthermore if we are going to make claims about this slowing down the kubelet it would be good to have some actionable metrics (which contrary to spam can be collected) rather than just wild guesses.

> Given these open questions, I think it is better address the spam detection/prevention problem in a separate PR.

Spam detection is not the scope of my argument. Spam prevention and keeping the little we have is.
I actually think it's a pretty big deal to loose the current behavior.
And given the rate at which PRs are getting merged in 1.9 (see the trivial PR of closing the endpoint connection) I do think we should use this opportunity to take a step in the right direction.

I don't see any real open questions though.

**Comment by jiayingz** [comment]
@RenaudWasTaken My main concern of including your suggested change in this PR is that the way it uses callback feels a bit tricky. I think the main motivation for this PR is to simplify the listwatch logic. If we still need to keep endpoint.listAndWatch() and wraps it with endpoint.run() as well as changing callback functions and adding timeout, it seems we make code even more complicated. Also one corner case the current code doesn't handle correctly is that we could fail to remove some devices that are no longer advertised by a re-registered device plugin: https://github.com/kubernetes/kubernetes/issues/55179. To solve this problem, one approach I can think of is to pass previously registered devices to the newly created endpoint. With that, there would be no callback issued if the re-registered device plugin advertises the same set of devices as its previous instance, and the callback trick you suggested would not work. I agree with you that we would like to prevent a malfunctional device plugin from flooding kubelet. However, I feel we may need more thinking on how to do this and don't want to block this PR with that.

**Comment by lichuqiang** [comment]
@jiayingz a commit is added in the PR to fix #55179

**Comment by jiayingz** [comment]
/lgtm

Given that we have different opinions on this PR, @vishh and @mindprince could you help review this PR and let us know your thoughts? Thanks!

File: pkg/kubelet/cm/deviceplugin/manager.go
Comment (mindprince) [inline pkg/kubelet/cm/deviceplugin/manager.go]: There is a `newEndpoint()` call before we `Unlock()` this `mutex`. The `newEndpoint()` call does a gRPC dial which can potentially take a long time. Can we avoid holding the lock for so long?

**Review by rohitagarwal003 — COMMENTED** [review]
LGTM in general except for one comment.

File: pkg/kubelet/cm/deviceplugin/manager.go
Comment (lichuqiang) [reply pkg/kubelet/cm/deviceplugin/manager.go]: Yes, I realized that too:
Here is related [comments](https://github.com/kubernetes/kubernetes/issues/55179#issuecomment-342453343) 
Do you have any suggestions?

Comment (jiayingz) [reply pkg/kubelet/cm/deviceplugin/manager.go]: Can we just use mutex to protect obtaining of old endpoint and its devices and adding new endpoint, but leaves newEndpoint out of the lock scope?

Comment (jiayingz) [reply pkg/kubelet/cm/deviceplugin/manager.go]: Looking at this a bit more, I realized some potential race condition with this change when we have two simultaneous endpoint creations, and also a small time window after we stop the old endpoint and before we set new endpoint. To close such races, can we modify the code like this:

existingDevs := make(map[string]pluginapi.Device)
m.mutex.Lock()
old, ok := m.endpoints[r.ResourceName]
if ok && old != nil {
   existingDevs = old.devices
}
m.mutex.Unlock()
socketPath := filepath.Join(m.socketdir, r.Endpoint)
e, err := newEndpoint(socketPath, r.ResourceName, existingDevs, m.callback)
if err != nil {
   glog.Errorf("Failed to dial device plugin with request %v: %v", r, err)
   return
}
m.mutex.Lock()
ext, ok := m.endpoints[r.ResourceName]
if ext != old {
  glog.Warningf("Some other endpoint %v is added while endpoint %v is initialized", ext, e)
  m.mutex.Unlock()
  e.stop()
  return
}
m.endpoints[r.ResourceName] = e
m.mutex.Unlock()
if old != nil {
  old.stop()
}

I don't quite like this code flow either because the logic is a bit complicated. I have been started to wonder whether we should simplify the callback interface between endpoint and manager, and make endpoint totally stateless. I.e., after we merge device plugin handler and manager, whenever endpoint receives a ListWatch rpc, it just sends the whole list of devices to device plugin manager, and manager uses that list to update its internal state stored in allDevices. My only concern with that approach is that a bad device plugin can keep sending redundant device updates that causes performance issue of manager as it needs to hold its lock to compare&update device states. My current feeling for this scenario is that we perhaps want to detect and prevent such spam in a separate mechanism outside the endpoint life cycle, and simplify the interface and implementation of endpoint.

Comment (lichuqiang) [reply pkg/kubelet/cm/deviceplugin/manager.go]: Oh, I tried to construct the code like you suggested too, but felt the structure a little unbearable somehow.
The idea to reconsider the responsibility of manager and endpoint sounds great. Also, I noticed that you have already removed `updateCapacityFunc` in callback, and extended device plugin checkpoint data to record registered resources in #55088, so I think the change wouldn't be complicated.
As for the protection mechanism, I would suggest we update the objects in `allDevices`  stateful, maybe with a timestamps to record last updated time, to ensure minimum time interval for update.

Comment (jiayingz) [reply pkg/kubelet/cm/deviceplugin/manager.go]: In the scope of this PR, perhaps we can use the mentioned code structure? I think the logic can be simplified later when we introduce a separate component to store state in manager.

Comment (lichuqiang) [reply pkg/kubelet/cm/deviceplugin/manager.go]: Updated with implementation above :)

File: pkg/kubelet/cm/deviceplugin/manager.go
Comment (tengqm) [inline pkg/kubelet/cm/deviceplugin/manager.go]: this looks ugly to me ... it could have been implemented differently, using a single, short lock while avoiding reregistration:

```
func (m *ManagerImpl) addEndpoint(r *pluginapi.RegisterRequest) {

    // The call involves an initial dial up which may take time to conclude
    e, err := newEndpoint(m, r)
    if err != nil {
        glog.Errorf("Failed to initialize endpoint %v: %v", r.Endpoint, err)
        return
    }

    // Now we lock the endpoints for mutation
    registered := false
    m.mutex.Lock()
    curr, ok := m.endpoints[r.ResourceName]
    if ok & curr != nil {
       registered = true
    } else {
       m.endpoints[r.ResourceName] = e
    }
    m.mutex.Unlock()

    if registered {
       e.stop()
       e = curr
    } else {
       curr.stop()
    }

    go func() {
       e.run()
       ...
    }()
}
```

In pkg/kubelet/cm/deviceplugin/endpoint.go:

```
func newEndpoint(m *Manager, resourceName string, r *pluginapi.RegisterRequest) (*endpoint, error) {
    socketPath := filepath.Join(m.SocketPath(), r.Endpoint)
    client, c, err := dial(socketPath)
    if err != nil {
        return nil, err
    }
    return &endpoint{
       client: client,
       clientConn: c,
       socketPath: socketPath,
       resourceName: resourceName,
       callback: m.callback,
    }  
}
```

Comment (lichuqiang) [reply pkg/kubelet/cm/deviceplugin/manager.go]: Well, the issue of re-registration is raised by the original purpose of passing existing devs into `newEndpoint` , plz take a look at #55179 and comments above

File: pkg/kubelet/cm/deviceplugin/endpoint.go
Comment (lichuqiang) [reply pkg/kubelet/cm/deviceplugin/endpoint.go]: the cleanup task is done in [manage layer](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/deviceplugin/manager.go#L243)

File: pkg/kubelet/cm/deviceplugin/endpoint.go
Comment (lichuqiang) [reply pkg/kubelet/cm/deviceplugin/endpoint.go]: the `listAndWatch` here indicates that in plugin client, so I think it ok to keep it as is.

File: pkg/kubelet/cm/deviceplugin/manager.go
Comment (RenaudWasTaken) [inline pkg/kubelet/cm/deviceplugin/manager.go]: This should not be in the lock

File: pkg/kubelet/cm/deviceplugin/manager.go
Comment (tengqm) [reply pkg/kubelet/cm/deviceplugin/manager.go]: well, I will check that PR review history. At the same time, I would suggest you to try review the logic I suggested above and see if it help simplify the whole function. It would also be a help to me so I know if I'm misunderstanding something.

**Comment by lichuqiang** [comment]
@jiayingz Sorry, I think I had a mistake when trying to update the UT for the re-registration corner scenario.
As `manager.Devices()` is that of the new endpoint, like @ScorpioCPH implied, the test would always succeed.
A e2e test case might be needed to reproduce the scenario.

**Comment by jiayingz** [comment]
I see the problem. I think we can still reproduce the scenario in this unit test. Actually I have been thinking that we need a follow-up PR after this one to clean up this unit test. First, I think we can now remove all the sleeps used in this test. Second, I think we can maintain a map of devices and updates it in the callback function based on added/removed devices reported by endpoints. Then we can check that the old device is correctly removed from the map after the last callback.

**Comment by lichuqiang** [comment]
@jiayingz Yes, I see, a store like a handler store stub would take effect.
Have you set about on it ? If not, I would like to work on the fix.

**Pull request merged**
